src/update_agent/actor.rs

```diff
@@ -463,13 +460,49 @@ impl UpdateAgentInfo {
         depls
     }
 
-    /// Return the update stream of the pending deployment.
-    async fn query_pending_deployment_stream(&self) -> Result<String> {
-        let msg = rpm_ostree::QueryPendingDeploymentStream {};
-        self.rpm_ostree_actor
+    /// Validate that pending deployment is coming from the correct update stream.
+    ///
+    /// This checks that the update stream for the pending deployment matches the current
+    /// one from agent identity.
+    /// The validation logic here is meant to run right after an update has been fetched,
+    /// thus it first ensures that a pending deployment exists with version and
+    /// checksum matching the one in input.
+    async fn is_pending_deployment_on_correct_stream(&self, release: Release) -> Result<bool> {
+        let msg = rpm_ostree::QueryPendingDeployment {
+            filter_hint: release.clone(),
+        };
+        let (pending, stream) = self
+            .rpm_ostree_actor
             .send(msg)
             .unwrap_or_else(|e| Err(e.into()))
-            .await
+            .await?;
+        if pending.version != release.version {
+            bail!("expected pending deployment '{}', but found '{}' instead");
+        }
+        if pending.checksum != release.checksum {
+            bail!(
+                "detected checksum mismatch for pending deployment '{}', got unexpected value '{}'",
+                release.version,
+                release.checksum,
+            );
+        }
+
+        if self.identity.stream != stream {
+            log::error!(
+                "pending deployment {} expected to be on stream '{}', but found '{}' instead",
+                pending.version,
+                self.identity.stream,
+                stream
+            );
+            Ok(false)
+        } else {
+            log::trace!(
+                "detected valid stream '{}' for pending deployment {}",
```

Comment:
"detected" has a slight negative connotation, and is used as such above in an error message.  Maybe:
```suggestion
                "validated stream '{}' for pending deployment {}",
```
?